File: scripts/zigux/check-phase8-file-path-handle-bridge-scope.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
from pathlib import Path


SCRIPT_PATH = Path(__file__).resolve()
ROOT = SCRIPT_PATH.parents[2] if len(SCRIPT_PATH.parents) > 2 else SCRIPT_PATH.parent

HELPER_PATH = ROOT / "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig"
TEST_PATH = ROOT / "zigux/tests/phase8_file_path_handle_bridge.zig"
MANIFEST_PATH = ROOT / "tools/lib/bpf/zigux_segments/manifest.json"

FDINFO_REQUIRED_HELPERS = [
    "pub fn buildProcFdinfoPath",
    "pub fn parseFdinfoLine",
    "pub fn applyFdinfoMapInfoLine",
    "pub fn parseFdinfoMapInfo",
    "pub fn summarizeFdinfoMapInfo",
]

FDINFO_REQUIRED_TEST_MARKERS = [
    "buildProcFdinfoPath",
    "parseFdinfoMapInfo",
    "applyFdinfoMapInfoLine",
]

MAP_REUSE_REQUIRED_HELPERS = [
    "pub fn resolveReusedMapName",
    "pub fn summarizeMapReuseCompatibility",
    "pub fn isMapReuseCompatible",
]

MAP_REUSE_REQUIRED_TEST_MARKERS = [
    "resolveReusedMapName",
    "summarizeMapReuseCompatibility",
    "isMapReuseCompatible",
]


class CheckFailure(Exception):
    pass


def load_text(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise CheckFailure(f"missing required file: {path.as_posix()}") from exc


def require_markers(text: str, markers: list[str], label: str) -> None:
    missing = [marker for marker in markers if marker not in text]
    if missing:
        formatted = ", ".join(repr(marker) for marker in missing)
        raise CheckFailure(f"{label} is missing required markers: {formatted}")


def find_segment(manifest: dict, slug: str) -> dict:
    for segment in manifest.get("segments", []):
        if segment.get("slug") == slug:
            return segment
    raise CheckFailure(f"manifest is missing segment {slug!r}")


def check_scope(helper_text: str, test_text: str, manifest: dict) -> None:
    fdinfo_segment = find_segment(manifest, "fdinfo-map-info-helpers")
    map_reuse_segment = find_segment(manifest, "map-reuse-compatibility")

    if fdinfo_segment.get("status") == "starter_landed":
        require_markers(
            helper_text,
            FDINFO_REQUIRED_HELPERS,
            "file-path bridge helper for fdinfo-map-info-helpers",
        )
        require_markers(
            test_text,
            FDINFO_REQUIRED_TEST_MARKERS,
            "phase8_file_path_handle_bridge tests for fdinfo-map-info-helpers",
        )

    if map_reuse_segment.get("status") == "starter_landed":
        require_markers(
            helper_text,
            MAP_REUSE_REQUIRED_HELPERS,
            "file-path bridge helper for map-reuse-compatibility",
        )
        require_markers(
            test_text,
            MAP_REUSE_REQUIRED_TEST_MARKERS,
            "phase8_file_path_handle_bridge tests for map-reuse-compatibility",
        )


def run_repo_check(root: Path) -> None:
    helper_text = load_text(root / HELPER_PATH.relative_to(ROOT))
    test_text = load_text(root / TEST_PATH.relative_to(ROOT))
    manifest = json.loads(load_text(root / MANIFEST_PATH.relative_to(ROOT)))
    check_scope(helper_text, test_text, manifest)


PASSING_MANIFEST = {
    "segments": [
        {"slug": "fdinfo-map-info-helpers", "status": "starter_landed"},
        {"slug": "map-reuse-compatibility", "status": "deferred_high_risk"},
    ]
}

PASSING_HELPER = """
pub fn buildProcFdinfoPath() void {}
pub fn parseFdinfoLine() void {}
pub fn applyFdinfoMapInfoLine() void {}
pub fn parseFdinfoMapInfo() void {}
pub fn summarizeFdinfoMapInfo() void {}
"""

PASSING_TEST = """
test "fdinfo" {
    _ = buildProcFdinfoPath;
    _ = parseFdinfoMapInfo;
    _ = applyFdinfoMapInfoLine;
}
"""

FAILING_MANIFEST = {
    "segments": [
        {"slug": "fdinfo-map-info-helpers", "status": "starter_landed"},
        {"slug": "map-reuse-compatibility", "status": "starter_landed"},
    ]
}

FAILING_HELPER = """
pub fn buildProcFdinfoPath() void {}
pub fn parseFdinfoLine() void {}
pub fn applyFdinfoMapInfoLine() void {}
pub fn parseFdinfoMapInfo() void {}
pub fn summarizeFdinfoMapInfo() void {}
"""

FAILING_TEST = """
test "fdinfo" {
    _ = buildProcFdinfoPath;
    _ = parseFdinfoMapInfo;
    _ = applyFdinfoMapInfoLine;
}
"""

FAILING_FDINFO_HELPER = """
pub fn buildProcFdinfoPath() void {}
pub fn applyFdinfoMapInfoLine() void {}
pub fn parseFdinfoMapInfo() void {}
pub fn summarizeFdinfoMapInfo() void {}
"""


def expect_failure(
    helper_text: str,
    test_text: str,
    manifest: dict,
    expected_scope: str,
) -> None:
    try:
        check_scope(helper_text, test_text, manifest)
    except CheckFailure as exc:
        message = str(exc)
        if expected_scope not in message:
            raise CheckFailure(
                f"self-test expected a {expected_scope} failure, got: {message}"
            ) from exc
        return
    raise CheckFailure(f"self-test expected a {expected_scope} scope failure")


def run_self_test() -> None:
    check_scope(PASSING_HELPER, PASSING_TEST, PASSING_MANIFEST)
    expect_failure(
        FAILING_FDINFO_HELPER,
        PASSING_TEST,
        PASSING_MANIFEST,
        "fdinfo-map-info-helpers",
    )
    expect_failure(
        FAILING_HELPER,
        FAILING_TEST,
        FAILING_MANIFEST,
        "map-reuse-compatibility",
    )
    print("PHASE8_FILE_PATH_HANDLE_BRIDGE_SCOPE_SELF_TEST=pass")


def main() -> int:
    parser = argparse.ArgumentParser(
        description=(
            "Fail closed when the Phase 8 file-path bridge manifest claims "
            "landed helper scope that the helper or focused tests do not expose."
        )
    )
    parser.add_argument(
        "--root",
        type=Path,
        default=ROOT,
        help="repository root to inspect (defaults to the inferred repo root)",
    )
    parser.add_argument(
        "--self-test",
        action="store_true",
        help="run built-in coverage checks instead of inspecting a repo",
    )
    args = parser.parse_args()

    try:
        if args.self_test:
            run_self_test()
        else:
            run_repo_check(args.root.resolve())
            print("PHASE8_FILE_PATH_HANDLE_BRIDGE_SCOPE=pass")
    except CheckFailure as exc:
        print(f"PHASE8_FILE_PATH_HANDLE_BRIDGE_SCOPE=fail: {exc}")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase8_file_path_handle_bridge_scope.zig
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const live_pass_marker = "PHASE8_FILE_PATH_HANDLE_BRIDGE_SCOPE=pass";
pub const self_test_pass_marker = "PHASE8_FILE_PATH_HANDLE_BRIDGE_SCOPE_SELF_TEST=pass";

const FDINFO_REQUIRED_HELPERS = [_][]const u8{
    "pub fn buildProcFdinfoPath",
    "pub fn parseFdinfoLine",
    "pub fn applyFdinfoMapInfoLine",
    "pub fn parseFdinfoMapInfo",
    "pub fn summarizeFdinfoMapInfo",
};

const FDINFO_REQUIRED_TEST_MARKERS = [_][]const u8{
    "buildProcFdinfoPath",
    "parseFdinfoMapInfo",
    "applyFdinfoMapInfoLine",
};

const MAP_REUSE_REQUIRED_HELPERS = [_][]const u8{
    "pub fn resolveReusedMapName",
    "pub fn summarizeMapReuseCompatibility",
    "pub fn isMapReuseCompatible",
};

const MAP_REUSE_REQUIRED_TEST_MARKERS = [_][]const u8{
    "resolveReusedMapName",
    "summarizeMapReuseCompatibility",
    "isMapReuseCompatible",
};

const PASSING_HELPER = [_][]const u8{
    "\npub fn buildProcFdinfoPath() void {}\npub fn parseFdinfoLine() void {}\npub fn applyFdinfoMapInfoLine() void {}\npub fn parseFdinfoMapInfo() void {}\npub fn summarizeFdinfoMapInfo() void {}\n",
};

const PASSING_TEST = [_][]const u8{
    "\ntest \"fdinfo\" {\n    _ = buildProcFdinfoPath;\n    _ = parseFdinfoMapInfo;\n    _ = applyFdinfoMapInfoLine;\n}\n",
};

const FAILING_HELPER = [_][]const u8{
    "\npub fn buildProcFdinfoPath() void {}\npub fn parseFdinfoLine() void {}\npub fn applyFdinfoMapInfoLine() void {}\npub fn parseFdinfoMapInfo() void {}\npub fn summarizeFdinfoMapInfo() void {}\n",
};

const FAILING_TEST = [_][]const u8{
    "\ntest \"fdinfo\" {\n    _ = buildProcFdinfoPath;\n    _ = parseFdinfoMapInfo;\n    _ = applyFdinfoMapInfoLine;\n}\n",
};

const FAILING_FDINFO_HELPER = [_][]const u8{
    "\npub fn buildProcFdinfoPath() void {}\npub fn applyFdinfoMapInfoLine() void {}\npub fn parseFdinfoMapInfo() void {}\npub fn summarizeFdinfoMapInfo() void {}\n",
};

fn checkRepo(io: Io, allocator: std.mem.Allocator, root: []const u8) !void {
    const text_fdinfo_required_helpers_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_fdinfo_required_helpers_path);
    const text_fdinfo_required_helpers = try guard.readUtf8File(io, allocator, text_fdinfo_required_helpers_path);
    defer allocator.free(text_fdinfo_required_helpers);
    for (FDINFO_REQUIRED_HELPERS) |marker| try guard.requireMarker(text_fdinfo_required_helpers, marker);
    const text_fdinfo_required_test_markers_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_fdinfo_required_test_markers_path);
    const text_fdinfo_required_test_markers = try guard.readUtf8File(io, allocator, text_fdinfo_required_test_markers_path);
    defer allocator.free(text_fdinfo_required_test_markers);
    for (FDINFO_REQUIRED_TEST_MARKERS) |marker| try guard.requireMarker(text_fdinfo_required_test_markers, marker);
    const text_map_reuse_required_helpers_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_map_reuse_required_helpers_path);
    const text_map_reuse_required_helpers = try guard.readUtf8File(io, allocator, text_map_reuse_required_helpers_path);
    defer allocator.free(text_map_reuse_required_helpers);
    for (MAP_REUSE_REQUIRED_HELPERS) |marker| try guard.requireMarker(text_map_reuse_required_helpers, marker);
    const text_map_reuse_required_test_markers_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_map_reuse_required_test_markers_path);
    const text_map_reuse_required_test_markers = try guard.readUtf8File(io, allocator, text_map_reuse_required_test_markers_path);
    defer allocator.free(text_map_reuse_required_test_markers);
    for (MAP_REUSE_REQUIRED_TEST_MARKERS) |marker| try guard.requireMarker(text_map_reuse_required_test_markers, marker);
    const text_passing_helper_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_passing_helper_path);
    const text_passing_helper = try guard.readUtf8File(io, allocator, text_passing_helper_path);
    defer allocator.free(text_passing_helper);
    for (PASSING_HELPER) |marker| try guard.requireMarker(text_passing_helper, marker);
    const text_passing_test_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_passing_test_path);
    const text_passing_test = try guard.readUtf8File(io, allocator, text_passing_test_path);
    defer allocator.free(text_passing_test);
    for (PASSING_TEST) |marker| try guard.requireMarker(text_passing_test, marker);
    const text_failing_helper_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_failing_helper_path);
    const text_failing_helper = try guard.readUtf8File(io, allocator, text_failing_helper_path);
    defer allocator.free(text_failing_helper);
    for (FAILING_HELPER) |marker| try guard.requireMarker(text_failing_helper, marker);
    const text_failing_test_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_failing_test_path);
    const text_failing_test = try guard.readUtf8File(io, allocator, text_failing_test_path);
    defer allocator.free(text_failing_test);
    for (FAILING_TEST) |marker| try guard.requireMarker(text_failing_test, marker);
    const text_failing_fdinfo_helper_path = try guard.joinPath(allocator, root, "tools/lib/bpf/zigux_segments/file_path_handle_bridge.zig");
    defer allocator.free(text_failing_fdinfo_helper_path);
    const text_failing_fdinfo_helper = try guard.readUtf8File(io, allocator, text_failing_fdinfo_helper_path);
    defer allocator.free(text_failing_fdinfo_helper);
    for (FAILING_FDINFO_HELPER) |marker| try guard.requireMarker(text_failing_fdinfo_helper, marker);
}

fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    try checkRepo(io, allocator, try guard.defaultRepoRoot(allocator));
    try guard.printLine(io, "{s}", .{self_test_pass_marker});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);

    var self_test = false;
    var explicit_root: ?[]const u8 = null;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    const root = explicit_root orelse try guard.repoRootFromScript(allocator);
    defer if (explicit_root == null) allocator.free(root);

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    checkRepo(io, allocator, root) catch {
        std.process.exit(1);
    };
    try guard.printLine(io, "{s}", .{live_pass_marker});
}
